torch/optim/optimizer.py

```diff
@@ -179,6 +179,7 @@ def zero_grad(self, set_to_none: bool = False):
                 (in one case it does the step with a gradient of 0 and in the other it skips
                 the step altogether).
         """
+        per_device_and_dtype_grads = defaultdict(lambda: defaultdict(list))
```

Comment:
Thanks for the excellent work on multi-tensor support @izdeby.  It really helps the performance a lot on internal production flows. Just one question I am discussing with @ngimel is that whether we could pre-compute the list or provide an option to do that, since the CPU spends quite a few time constructing the list.

Reply:
Glad you're trying it, and glad you're alert to the issue.  I showed Iurii how to use nsight systems to assess CPU and GPU overhead (https://gist.github.com/mcarilli/376821aa1a7182dfcf59928a7cde3223).  We'll figure out how to reduce it over time (ie more efficient Python expressions).  It's tricky because the set of params that receive a grad may change across iterations.

Reply:
Yes, that's true in general. What if we don't have dynamic path in the graph (actually most cases), or even applying extra zero grad does not change the parameter/state result(it might depend on concrete optimizer algorithm).

Reply:
@lly-zero-one, sorry for the late response. Yeah, CPU perf is something we are thinking about and will definitely take a close look at it sooner rather than later.